syntax = "proto3";

package raft;

message LogEntry {
    bool noop = 1;
    uint64 term = 2;
    uint64 index = 3;
    bytes entry = 4;
}

message AppendEntriesRequest {
    uint64 term = 1;
    uint64 leaderId = 2;

    uint64 prevLogIndex = 3;
    uint64 prevLogTerm = 4;
    repeated LogEntry entries = 5;
    uint64 leaderCommit = 6;
    uint64 peer_id = 7;
}

message AppendEntriesResponse {
    uint64 term = 1;
    bool success = 2;
    uint64 peer_id = 3;

    uint64 request_term = 4;
    uint64 match_index = 5;
    uint64 last_log_index = 6;
}

message RequestVoteRequest {
    uint64 term = 1;
    uint64 candidateId = 2;
    uint64 lastLogIndex = 3;
    uint64 lastLogTerm = 4;
    uint64 peer_id = 5;
}

message RequestVoteResponse {
    uint64 term = 1;
    bool voteGranted = 2;
    uint64 peer_id = 3;
}

service RaftService {
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse) {}
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse) {}
}
